### PR TITLE
Allow to listen on multiple IP addresses

### PR DESCRIPTION
**Pull request by smortex**
Allow to pass multiple IP addresses.  The default behavior is to listen on IPv4 only so this is only a breaking change because the interface change.

### CODE REVIEW

**Comment by smortex** [comment]
But maybe since #173 breaks backward compatibility, no need to support a `String` for the parameters and only accept `Array[String[1]]`?

**Review by neomilium — APPROVED** [review]

**Comment by neomilium** [comment]
> But maybe since #173 breaks backward compatibility, no need to support a `String` for the parameters and only accept `Array[String[1]]`?

I agree, this makes code more simple.

**Comment by smortex** [comment]
:ballot_box_with_check:  backward compatibility code removed

**Review by neomilium — APPROVED** [review]

**Comment by neomilium** [comment]
@smortex Whould you like I merge too on this project after a review?

**Comment by smortex** [comment]
I would prefer to have a second approval by somebody external of @opus-codium we both belong to before merging :wink: 

If we have no news in a few day, no objection to pull the trigger.

File: manifests/client.pp
Comment (smortex) [inline manifests/client.pp:40]: `['0.0.0.0', '::']` might make more sense… Opinion?

Comment (smortex) [reply manifests/client.pp:40]: Maybe the most logical is to default to `[]`: do not output configuration and behave the way bacula developers chose it should work out-of-the-box?

Comment (bastelfreak) [reply manifests/client.pp:40]: if bacula has a proper default builtin, I think using that is a good idea.

Comment (neomilium) [reply manifests/client.pp:40]: IMHO, no defaults, ie. required param, is better than `[]`.

If defaults is to listen on any interface, we will have to use some hiera settings per OS to workaround the `::` OS-specific behavior.

Comment (smortex) [reply manifests/client.pp]: The default is equivalent to what we had, so let's use the default.

**Comment by smortex** [comment]
I updated the PR to not set an explicit listen address by default, which bacula understands as listen on all IPv4 only (so same behavior as before).

**Review by neomilium — CHANGES_REQUESTED** [review]
> I updated the PR to not set an explicit listen address by default, which bacula understands as listen on all IPv4 only (so same behavior as before).

IMHO, this behavior should be shorty documented. Could you put a note in the variable documentation?

**Comment by neomilium** [comment]
> The only thing that annoys me here is that the service only listen on IPv4 by default (which is an issue at Bacula's level), and I do not think the module documentation is the right place to describe this since this can change at any point upstream?

OK, so we had to support OS-independent and idempotent way. What about putting some hiera files, OS-dependant to reach the same behavior on each OS?

**Comment by smortex** [comment]
> > The only thing that annoys me here is that the service only listen on IPv4 by default (which is an issue at Bacula's level), and I do not think the module documentation is the right place to describe this since this can change at any point upstream?
> 
> OK, so we had to support OS-independent and idempotent way. What about putting some hiera files, OS-dependant to reach the same behavior on each OS?

I really don't know how we could deal with this: system dependent config also look wrong to me: the default behavior of FreeBSD and e.g. Debian is not the same in the way that Debian offer IPv6 mapped IPv4 by default while FreeBSD doesn't. But as stated above, Debian can be tuned to opt-out, and of course FreeBSD [can be tuned to opt-in](https://docs.freebsd.org/en/books/handbook/advanced-networking/#_ipv6_and_ipv6_address_mapping) :shrug: Above I wrote "Linux" but I really meant "Debian", I don't know if the default is the same for every Linux distro, and I would not be surprised to see it is not. And since it is tunable, I think that sticking to upstream default is the best way to not break POLA.

Maybe in the end a note in the README is better?  What information would you like to see there?

**Comment by smortex** [comment]
> Could you put a note in the variable documentation?

"variable documentation" :exploding_head: 

Yeah!  Did that! Thanks!

**Review by neomilium — APPROVED** [review]
Excellent explanation!

**Pull request merged**
